### ISSUE TITLE
Can not replicate paper's results

### ISSUE THREAD

**Issue by YESAndy** [wz0919/waypoint-predictor#2]
Hi,

Thank you very much for sharing this great work! I successfully retrained the waypoint predictor but found that my result was not good as the paper had shown. My results are shown below:

Epoch: 300/300 |██████████| 100.0% 
Train Loss: 14.47815
Eval Loss: 14.26238
Epoch 299 p_waypoint_openspace 0.38272 | p_waypoint_obstacle 0.61728 | avg_wayscore 0.35390 | avg_pred_distance 1.46896 | avg_chamfer_distance 1.30407 | avg_hausdorff_distance 2.42182 | avg_num_delta 0.76277 | 
Best avg_wayscore result til now:  Epoch 47 p_waypoint_openspace 0.35387 | p_waypoint_obstacle 0.64613 | avg_wayscore 0.43396 | avg_pred_distance 1.24561 | avg_chamfer_distance 1.25932 | avg_hausdorff_distance 2.32100 | avg_num_delta 1.65239 | 
Best avg_pred_distance result til now:  Epoch 8 p_waypoint_openspace 0.38868 | p_waypoint_obstacle 0.61132 | avg_wayscore 0.38347 | avg_pred_distance 0.96149 | avg_chamfer_distance 1.05154 | avg_hausdorff_distance 2.04617 | avg_num_delta 1.34349 | 

I can think of one reason that I collected the training split (61) and val_unseen (11) split from [vln_r2r_mp3d_v1](https://dl.fbaipublicfiles.com/habitat/data/datasets/vln/mp3d/r2r/v1/vln_r2r_mp3d_v1.zip) because I cannot find train.json as mentioned in this repo. Can you provide the train.json?

Many Thankkkkkssss!

**Comment by wz0919** [comment]
RAW_GRAPH_FILE https://drive.google.com/drive/folders/1wpuGAO-rRalPKt8m1-QIvlb_Pv1rYJ4x?usp=sharing

**Comment by chen-judge** [comment]
I directly tested the model checkpoint (_check_val_best_avg_wayscore_fov90_rgbd_) released in this repo: [https://github.com/YicongHong/Discrete-Continuous-VLN](https://github.com/YicongHong/Discrete-Continuous-VLN) and obtained the following results:
`Epoch 152 p_waypoint_openspace 0.43507 | p_waypoint_obstacle 0.56493 | avg_wayscore 0.39565 | avg_pred_distance 1.29666 | avg_chamfer_distance 1.21823 | avg_hausdorff_distance 2.27901 | avg_num_delta 1.42834 |`


However, I found that these results seem to be worse than those mentioned in the paper. I have alread used the RAW_GRAPH_FILE you provided, to set the RAW_GRAPH_PATH in _test_twm0.2_obstacle_first.py_. 

What other possible reasons could have led to these results?

Looking forward to your reply. Thanks!!

**Comment by wz0919** [comment]
> I directly tested the model checkpoint (_check_val_best_avg_wayscore_fov90_rgbd_) released in this repo: https://github.com/YicongHong/Discrete-Continuous-VLN and obtained the following results: `Epoch 152 p_waypoint_openspace 0.43507 | p_waypoint_obstacle 0.56493 | avg_wayscore 0.39565 | avg_pred_distance 1.29666 | avg_chamfer_distance 1.21823 | avg_hausdorff_distance 2.27901 | avg_num_delta 1.42834 |`
> 
> However, I found that these results seem to be worse than those mentioned in the paper. I have alread used the RAW_GRAPH_FILE you provided, to set the RAW_GRAPH_PATH in _test_twm0.2_obstacle_first.py_.
> 
> What other possible reasons could have led to these results?
> 
> Looking forward to your reply. Thanks!!

Hi,

I just reran the code and found a typo in `get_images_inputs.py`. `NUMBER` should be set to 12 instead of 120. The bad performance may be caused by using the first 12 images out of the whole 120 under `NUMBER=120` settings. I'm sorry for this typo and  I've fixed it.
Then, if there're no modifications to the codes and environments, you should get this result (which I got just now).
`Epoch 152 p_waypoint_openspace 0.79865 | p_waypoint_obstacle 0.20135 | avg_wayscore 1.42925 | avg_pred_distance 1.01193 | avg_chamfer_distance 1.07161 | avg_hausdorff_distance 2.00423 | avg_num_delta 1.39868 |`

**Comment by chen-judge** [comment]
Thank you for your reply!

What about the `NUMBER` in the get_nav_dict.py: [https://github.com/wz0919/waypoint-predictor/blob/ac357545c1a768e11c41c1686f35deb82462ecbb/gen_training_data/get_nav_dict.py#L11](https://github.com/wz0919/waypoint-predictor/blob/ac357545c1a768e11c41c1686f35deb82462ecbb/gen_training_data/get_nav_dict.py#L11)

Should I modify this as well?

**Comment by wz0919** [comment]
No. NUMBER in get_images_inputs.py relates to the number of input panoramic images (default 12 images in 360 degrees so it should be 12), while NUMBER in get_nav_dict.py relates to how fine-grained the predicted heatmap is (NUMBER=120 means your heatmap data will be 120x12 grid, which we used in our experiments).

Seems I modified get_nav_dict.py to get get_image_inputs.py (So they look similar) and forgot to modify NUMBER.

**Comment by chen-judge** [comment]
Thank you for your prompt reply and help! I've already reproduced the results.

BTW, do you know how to visualize the waypoint prediction in visual observations? For example, the left image in this GIF :
[https://github.com/YicongHong/Discrete-Continuous-VLN/blob/main/figures/traj_0.gif](https://github.com/YicongHong/Discrete-Continuous-VLN/blob/main/figures/traj_0.gif)

I have found a function _draw_waypoint_prediction_ that can plot global coordinates on a top-down map, but how can I  draw these coordinates in RGB observations?

```Python
maps.draw_waypoint_prediction(
            top_down_map,
            _rtheta_to_global_coordinates(
                action["action_args"]["r"],
                action["action_args"]["theta"],
                start_pos,
                heading_from_quaternion(start_heading),
            ),
            info[map_k]["meters_per_px"],
            info[map_k]["bounds"],
        )
```

Thank you in advance!

**Comment by wz0919** [comment]
It's a bit hard to provide the drawing code as it's extremely messy.

In habitat you can insert objects in the environments. We add cylinders to be our waypoints. You can also remove them. Tutorials are from [here](https://aihabitat.org/docs/habitat-sim/rigid-object-tutorial.html). Here's a raw code we used which may be helpful (we use it as a function in environments.py).
```
# set sylinder
obj_attr_mgr = self._env.sim.get_object_template_manager()
self.sphere_handle = obj_attr_mgr.get_template_handles("cylinderSolid")[0]
# red
lighting_setup_1 = [
    LightInfo(vector=[0.0, 0.0, 1.0, 1.0],color=[200.,0.,0.],model=LightPositionModel.CAMERA)
]
self._env.sim.set_light_setup(lighting_setup_1, "choosed_wp")
# blue
lighting_setup_2 = [
    LightInfo(vector=[0.0, 0.0, 1.0, 1.0],color=[0.,0.,200.],model=LightPositionModel.CAMERA)
]
self._env.sim.set_light_setup(lighting_setup_2, "other_wp")
# snap waypoint in obstacles
point = self.habitat_env.sim.pathfinder.snap_point(point)
# add sylinder
cp_id =  self._env.sim.add_object_by_handle(self.sphere_handle)
self._env.sim.set_translation(point, cp_id)
# set light
self._env.sim.set_object_light_setup(cp_id,wp,LightPositionModel.CAMERA)
```
BTW, Jacob has a very nice visualization codebase in [sim2sim transfer](https://github.com/jacobkrantz/Sim2Sim-VLNCE) which may be very helpful too.

**Comment by chen-judge** [comment]
Great! Thank you for your patient response and valuable advice!!!

**Comment by hover03** [comment]
@chen-judge hi, i wander if you would like to share the weight file "check_val_best_avg_wayscore_fov90_rgbd", since the link has become invalid in the repo, thanks a lot!!!
